proposals/extensions/android-parcelable.md

```diff
@@ -0,0 +1,263 @@
+                val data = source.readInt()
+                return MyParcelable(data)
+            }
+
+            override fun newArray(size: Int) = arrayOfNulls<MyParcelable>(size)
+        }
+    }
+}
+```
+
+This creates even more nesting and requires even more details to be memorized. There's a community [plugin](https://github.com/nekocode/android-parcelable-intellij-plugin-kotlin) for IntelliJ IDEA/Android Studio that generates this boilerplate. The [PaperParcel] library seems to work with Kotlin too, but requires manual creation of `CREATOR` fields:
+
+```kotlin
+@PaperParcel
+data class User(
+    val id: Long,
+    val firstName: String,
+    val lastName: String
+) : PaperParcelable {
+  companion object {
+    @JvmField val CREATOR = PaperParcelUser.CREATOR
+  }
+}
+```
+
+## Proposal
+
+We can have a compiler extension to generate all the required boilerplate behind the scenes. This is similar to what (some) annotation processors do, but 
+1. Java annotation processors can't alter Kotlin code, 
+2. Normally, annotation processors can't add methods to existing classes (this is possible though a private API only),
+3. A compiler extension can be potentially more flexible in terms of suppressing errors and providing syntactic means to the user.  
+
+### Simple case: completely automatic Parcelable
+
+A compiler extension can generate serialization/deserialization logic for all properties in a primary constructor of a class marked with a special annotation:
+
+```kotlin
+@Parcelize
+class MyParcelable(val data: Int): Parcelable
+```  
+
+Note that the class is not required to be a data class.
+
+The following requirements apply here:
+- Only non-`abstract` classes are supported
+  - `@Parcelize` can't annotate an interface
+  - `sealed` and `enum` classes may be supported
+  - `object` are forbidden for now
+    - Objects can also be supported. We can serialize just the qualified name because we can't make the new instance of `object` anyway
+- The class annotated with `@Parcelize` must implement `Parcelable` (directly or through a chain of supertypes)
+  - Otherwise it's a compiler error
+- All parameters of the primary constructor must be properties, otherwise they can not be (de)serialized
+  - If there is a non-property parameter, it is a compilation error
+- Properties with initializers declared in the class body are also difficult to deserialize correctly: we'd have to generate an alternative constructor that does not execute initializers (including `init` blocks) at all and only uses the serialized data, but this has all the issues that `java.io.Serializable` has wrt "magic" object creation.
+  - Properties in the class body must be marked with a `@Transient` annotation, otherwise it's a compiler warning (or error?) 
+- If some properties are not parcelable (i.e. do not implement `Parcelable`, are not supported by the `Parcel` interface, and are not customized (see below)), it's a compilation error
```

Comment:
nit: assume this means properties appearing in the constructor, but it's not stated explicitly.